core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -4054,12 +4080,12 @@ object ZIO extends ZIOCompanionPlatformSpecific with ZIOCompanionVersionSpecific
     initial: => S
   )(cont: S => Boolean, inc: S => S)(body: S => ZIO[R, E, Any])(implicit trace: Trace): ZIO[R, E, Unit] =
     ZIO.suspendSucceed {
-
-      def loop(initial: S): ZIO[R, E, Unit] =
-        if (cont(initial)) body(initial) *> loop(inc(initial))
-        else Exit.unit
-
-      loop(initial)
+      var s = initial
+      ZIO.whileLoop(cont(s)) {
+        val effect = body(s)
+        s = inc(s)
+        effect
+      }(_ => ())
```

Comment:
Nit: Use `ZIO.unitFn` instead of `_ => ()`. Also noticed this in a couple of other places in this PR, so maybe make that change there as well